presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java

```diff
@@ -2113,6 +2125,65 @@ private void verifySelectDistinct(QuerySpecification node, List<Expression> outp
             return orderByFields;
         }
 
+        private void analyzeLimit(Query node)
+        {
+            checkState(node.getLimit().isPresent(), "Fetch first or limit is absent");
+            analyzeLimit(node, node.getLimit().get());
+        }
+
+        private void analyzeLimit(QuerySpecification node)
+        {
+            checkState(node.getLimit().isPresent(), "Fetch first or limit is absent");
+            analyzeLimit(node, node.getLimit().get());
+        }
+
+        private void analyzeLimit(Node node, Node fetchFirstOrLimit)
+        {
+            checkState(fetchFirstOrLimit instanceof FetchFirst || fetchFirstOrLimit instanceof Limit, "Invalid limit node type. Expected: FetchFirst or Limit. Actual: %s", fetchFirstOrLimit.getClass());
+            if (fetchFirstOrLimit instanceof FetchFirst) {
+                analyzeLimit(node, (FetchFirst) fetchFirstOrLimit);
+            }
+            else {
+                analyzeLimit(node, (Limit) fetchFirstOrLimit);
+            }
+        }
+
+        private void analyzeLimit(Node node, FetchFirst fetchFirst)
+        {
+            if (!fetchFirst.getRowCount().isPresent()) {
+                analysis.setLimit(node, 1L);
+            }
+            else {
+                long rowCount;
+                try {
+                    rowCount = Long.parseLong(fetchFirst.getRowCount().get());
+                }
+                catch (NumberFormatException e) {
+                    throw new SemanticException(INVALID_FETCH_FIRST_ROW_COUNT, node, "%s", fetchFirst.getRowCount().get());
```

Comment:
The node in the exceptions should be the `fetchFirst` (or `limit` in the next method). That's used for generating the appropriate location in the error message.